site/content/entry/yuzu-progress-report-may-2022/index.md

```diff
@@ -0,0 +1,272 @@
+
+## Graphical changes, driver issues, and the nostalgia bliss that is the good old 64
+
+This month, [byte[]](https://github.com/liamwhite) continues the wave of improvements for `Super Mario 3D All-Stars`, this time, he noticed a bug in the DMAcopy ([direct memory access](https://en.wikipedia.org/wiki/Direct_memory_access)) of the Nintendo Switch’s GPU.
+
+`DMACopy` is a mechanism that many games use to send texture data to the GPU, it handles the format conversion from "pitch" (pixels on a line by line basis) to "tiled" (gridded) images.
+This process works by writing the pitch image data into GPU memory accessible by the DMA engine. Next, a DMAcopy is requested through the DMA engine driver, converting the image data into a separate buffer accessible by the GPU. This buffer will then be used as the texture on the final draw.
+
+{{< gh-hovercard "8313" "After fixing `bytes_per_pixel`," >}} `Super Mario Galaxy` now has proper lens flare.
+
+{{< single-title-imgs-compare
+    "RTX On? (Super Mario Galaxy)"
+    "./dmabug.png"
+    "./dmafix.png"
+>}}
+
+byte[] also improved the way OpenGL interprets face flips depth, [replacing the previously reported fix](https://yuzu-emu.org/entry/yuzu-progress-report-apr-2022/#saving-princess-peach-yet-again). The previous implementation had terrible peformance in OpenGL, making it virtually unplayable.
+
+While this wasn't an issue while using Vulkan, now `Super Mario 64` and `Super Mario Galaxy` {{< gh-hovercard "8314" "are playable in both APIs." >}} 
+Fermi GPU users rejoyce.
+
+{{< imgs
+    "./sm64.png| It's-a Mario, now running fast in OpenGL! (Super Mario 64)"
+  >}}
+
+One of the important parts of yuzu's graphical emulation is the need to translate small sets of GPU instructions, called `macros`. 
+yuzu uses a Just-in-Time (JIT) compiler to emulate these macros in a performant way.
+MacroJIT is an optimization applied over this process, it manages to provide a performance boost of plus 10%, more or less.
+
+byte[] found that macro could try to access a parameter that was too far outside the bounds of what it was supposed to be accessing. 
+Usually the end result of this behaviour would be the emulator crashing.
+{{< gh-hovercard "8319" "One less reason for annoying crashes." >}}
+
+Now, while on the subject or macros, MME, or Macro Method Expander, are small programs sent to the GPU when booting a game, responsible for executing methods (anything that changes the current status of the GPU).
+byte[] added the option to {{< gh-hovercard "8320" "dump such MME macros" >}} for debugging purposes.
+
+But why are MME macros important enough to merit their own dump mechanism?
+
+Turns out, the `Nintendo 64` emulator (*totally not outside Nintendo's Terms of Service*), included with the `Nintendo Switch Online` subscription, reasigns the same macros multiple times, each with different code.
```

Comment:
```suggestion
Turns out, the `Nintendo 64` emulator (*totally not outside Nintendo's Terms of Service*), included with the `Nintendo Switch Online` subscription, reassigns the same macros multiple times, each time with different code. yuzu incorrectly appended the new code to the end of the macro in this case, instead of replacing the existing code.
```